Add a TypeScript type for Pulumi Schema

mikhailshilkov opened the pull request:
Adds TypeScript types to represent Pulumi Schema in our node SDK. It's needed to be able to generate a schema from a TypeScript program (the case in mind is "reflection"-based component schema generation).

Types aren't exported from the SDK, so they can stay internal for now.

Let me know if we have these types somewhere already, or if there is a better home for them.

justinvp commented:
How are these generated?

> Types aren't exported from the SDK, so they can stay internal for now.

I suspect they're going to show up in generated docs at https://www.pulumi.com/docs/reference/pkg/nodejs/pulumi/pulumi/ unless they are tagged with `/** @internal */`

mikhailshilkov commented:
> How are these generated?

We should consider them manually written. I tried three different tools + AI and all of them had trouble converting everything truthfully. In the edit, I manually edited one of those results. It's not easy to re-generate the types. But also, I don't think the schema changes that often?

> I suspect they're going to show up in generated docs

Maybe that's fine? I didn't export them because I don't know how users may use them, but maybe there is not much harm? If we find issues, we can always fix them without caring about breaking changes, because the JSON schema is the source of truth, right?

danielrbradley commented:
For reference, here's another spike I did a while back on generating types for all SDKs via defining our meta-schema in Pulumi Schema: https://github.com/pulumi/pulumi-schema

- Node Types: https://github.com/pulumi/pulumi-schema/blob/master/nodejs/types/output.ts
- Go Types: https://github.com/pulumi/pulumi-schema/blob/master/go/pulumiTypes.go

mikhailshilkov commented:
I took another look at automatic generation and was able to produce something decent with `quicktype` and some post-processing. Logging the script here for now, will have to ship it separately:

```js
const { quicktype, InputData, JSONSchemaInput, TypeScriptTargetLanguage } = require("quicktype-core");
const fs = require('fs/promises');

async function convertSchema() {
    const schema = await fs.readFile('pulumi.json', 'utf8');
    
    const inputData = new InputData();
    const schemaInput = new JSONSchemaInput(new FetchingJSONSchemaStore());
    await schemaInput.addSource({ name: "PulumiPackage", schema });
    inputData.addInput(schemaInput);

    const { lines } = await quicktype({
        inputData,
        lang: new TypeScriptTargetLanguage(),
        rendererOptions: {
            "just-types": "true",  // This removes the runtime code
            "explicit-unions": "true"
        }
    });

    // Post-process to convert enums to union types
    let content = lines.join('\n');
    const enumRegex = /export enum (\w+) \{([^}]+)\}/g;
    content = content.replace(enumRegex, (match, enumName, enumBody) => {
        const values = enumBody
            .split(',')
            .map(line => line.trim())
            .filter(line => line)
            .map(line => {
                const [key, value] = line.split('=').map(s => s.trim());
                return value;
            })
            .join(' | ');
        return `export type ${enumName} = ${values};`;
    });

    await fs.writeFile('types.ts', content);
}

class FetchingJSONSchemaStore {
    async fetch(address) {
        return undefined;
    }
}

convertSchema().catch(console.error);
```
